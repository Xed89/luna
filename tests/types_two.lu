type Program1
  fun main()
    Console.writeLine("Hello World!")

type Program2
  fun main()
    Console.writeLine("Hello World!")