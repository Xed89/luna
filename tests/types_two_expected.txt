Parsing completed
Syntax Tree:
TypeDeclaration 'Program1'
  [0]: 
    Function 'main'
      type:
        null
      statements:
        [0]: 
          VarOrCallChainMaybeAssignStatementSyntax
            VarOrCallChainSyntax
              [0]: 
                VariableOrCall 'Console'
              [1]: 
                VariableOrCall 'writeLine'
                  [0]: 
                    ExpressionLiteralSyntax 'Hello World!'
TypeDeclaration 'Program2'
  [0]: 
    Function 'main'
      type:
        null
      statements:
        [0]: 
          VarOrCallChainMaybeAssignStatementSyntax
            VarOrCallChainSyntax
              [0]: 
                VariableOrCall 'Console'
              [1]: 
                VariableOrCall 'writeLine'
                  [0]: 
                    ExpressionLiteralSyntax 'Hello World!'

Compiled module:
Module types_two
  Type Program1
    Function main()
      LunaCompiler.VarOrCallChainMaybeAssignStatement
  Type Program2
    Function main()
      LunaCompiler.VarOrCallChainMaybeAssignStatement

Generated Cpp code:
#include "stdio.h"
#include "stdarg.h"
class Console {
public:
  static void writeLine(const char* msg, ...) {
    va_list argp;
    va_start(argp, msg);
    vfprintf(stdout, msg, argp);
    fprintf(stdout, "\n");
    va_end(argp);
  }
};

class Program1 {
public:
  static void main() {
    Console::writeLine("Hello World!");
  }
};
class Program2 {
public:
  static void main() {
    Console::writeLine("Hello World!");
  }
};
int main() {
  Program1::main();
}

Program output:
Hello World!
